AX: Hover Text does not enlarge input values

twilco opened the pull request:
#### ca4a1990df62ce5cac295d1163cc447941e51756
<pre>
AX: Hover Text does not enlarge input values
<a href="https://bugs.webkit.org/show_bug.cgi?id=299130">https://bugs.webkit.org/show_bug.cgi?id=299130</a>
<a href="https://rdar.apple.com/160891414">rdar://160891414</a>

Reviewed by Joshua Hoffman.

The reason this bug happened was because when HoverText did a hit test on an input, we returned the deepest unignored
match at the point, which was a generic div rendered inside the user-agent shadow tree generated by input elements.
There is no reason to expose this div in the accessibility tree, so ignore it, in turn causing the hit test to return
the text input, which HoverText knows how to deal with.

* LayoutTests/accessibility/text-input-hit-test-expected.txt: Added.
* LayoutTests/accessibility/text-input-hit-test.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):

Canonical link: <a href="https://commits.webkit.org/300691@main">https://commits.webkit.org/300691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/72e6d9ec408a542a05e5b33214f0118b4b89fe01

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123429 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43144 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33840 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130145 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75562 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f1b7c746-d2ec-40cc-a992-938221193e0f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125306 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43867 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51738 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93819 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62265 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4e2cab7e-c5bf-4f2f-95e1-c273cdf1dc78) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126382 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34951 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110426 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74446 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d4670a45-8ad9-4638-b728-bae3c75f0335) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28584 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73663 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104665 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132865 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102311 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50755 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102163 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25988 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47522 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25747 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47184 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50234 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55995 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49706 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53055 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51383 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1430,8 +1431,14 @@ bool AccessibilityRenderObject::computeIsIgnored() const
-    if (isGenericFocusableElement() && node->firstChild())
+    if (isGenericFocusableElement() && node->firstChild()) {
+        RefPtr shadowRoot = node->containingShadowRoot();
+        if (shadowRoot && shadowRoot->mode() == ShadowRootMode::UserAgent && is<HTMLInputElement>(shadowRoot->host())) {
```
Is HTMLInputElement the only one that might have this issue? How about textarea, select, <video controls>, etc? I wonder if there's a more general check than casting to HTMLInputElement...

twilco replied:
Excellent question! I checked textarea, select, audio, and video, and all are fine. So I think just checking for HTMLInputElement is enough!

minorninth reviewed (commented):
lgtm with one question

twilco committed:
AX: Hover Text does not enlarge input values
https://bugs.webkit.org/show_bug.cgi?id=299130
rdar://160891414

Reviewed by Joshua Hoffman.

The reason this bug happened was because when HoverText did a hit test on an input, we returned the deepest unignored
match at the point, which was a generic div rendered inside the user-agent shadow tree generated by input elements.
There is no reason to expose this div in the accessibility tree, so ignore it, in turn causing the hit test to return
the text input, which HoverText knows how to deal with.

* LayoutTests/accessibility/text-input-hit-test-expected.txt: Added.
* LayoutTests/accessibility/text-input-hit-test.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):

Canonical link: https://commits.webkit.org/300691@main
